File: tests/testthat/test-messaging.R
test_that("two peers can discover each other via get_user", {
  skip_if_not(server_is_running(), "test server not available")

  # Connect Alice
  peer_connect("alice", "alice_pass")
  peer_receive(500L)  # INIT_RESPONSE
  peer_disconnect()

  # Connect Bob and discover Alice
  peer_connect("bob", "bob_pass")
  peer_receive(500L)  # INIT_RESPONSE

  expect_no_error(peer_get_user("alice", "alice_pass"))
  pkt <- peer_receive(1000L)

  # Should get START_PINGING_PEER (type 4)
  expect_equal(pkt, 4L)

  peer_disconnect()
})

test_that("peer can send and receive a message", {
  skip_if_not(server_is_running(), "test server not available")

  # Connect sender
  peer_connect("sender", "sender_pass")
  peer_receive(500L)

  # We need a receiver - but since our library is single-instance (static globals),

  # we can only run one client at a time. Test that send_message succeeds
  # when the peer is known (discovered via get_user).
  peer_disconnect()

  # Connect receiver first so it's registered
  peer_connect("receiver", "recv_pass")
  peer_receive(500L)
  peer_disconnect()

  # Connect sender and discover receiver
  peer_connect("sender2", "sender2_pass")
  peer_receive(500L)
  peer_get_user("receiver", "recv_pass")
  pkt <- peer_receive(1000L)
  expect_equal(pkt, 4L)  # START_PINGING_PEER

  # Send message to discovered peer (should not error)
  expect_no_error(peer_send_message("receiver", "Hello from test!"))

  peer_disconnect()
})

test_that("peer can ping a discovered peer", {
  skip_if_not(server_is_running(), "test server not available")

  # Register target
  peer_connect("ping_target", "pt_pass")
  peer_receive(500L)
  peer_disconnect()

  # Connect pinger and discover target
  peer_connect("pinger", "pinger_pass")
  peer_receive(500L)
  peer_get_user("ping_target", "pt_pass")
  pkt <- peer_receive(1000L)
  expect_equal(pkt, 4L)

  expect_no_error(peer_send_ping("ping_target"))

  peer_disconnect()
})

File: tests/pytest/test_messaging.py
import PeerChat


def test_discover_peer_via_get_user():
    # Register alice
    PeerChat.connect("py_alice", "alice_pass")
    PeerChat.receive(500)  # INIT_RESPONSE
    PeerChat.disconnect()

    # Connect bob and discover alice
    PeerChat.connect("py_bob", "bob_pass")
    PeerChat.receive(500)  # INIT_RESPONSE

    PeerChat.get_user("py_alice", "alice_pass")
    pkt = PeerChat.receive(1000)

    # START_PINGING_PEER = 4
    assert pkt == 4

    PeerChat.disconnect()


def test_send_message_to_discovered_peer():
    # Register receiver
    PeerChat.connect("py_receiver", "recv_pass")
    PeerChat.receive(500)
    PeerChat.disconnect()

    # Connect sender and discover receiver
    PeerChat.connect("py_sender", "sender_pass")
    PeerChat.receive(500)
    PeerChat.get_user("py_receiver", "recv_pass")
    pkt = PeerChat.receive(1000)
    assert pkt == 4

    # Should not raise
    PeerChat.send_message("py_receiver", "Hello from pytest!")

    PeerChat.disconnect()


def test_ping_discovered_peer():
    # Register target
    PeerChat.connect("py_ping_target", "pt_pass")
    PeerChat.receive(500)
    PeerChat.disconnect()

    # Connect pinger and discover target
    PeerChat.connect("py_pinger", "pinger_pass")
    PeerChat.receive(500)
    PeerChat.get_user("py_ping_target", "pt_pass")
    pkt = PeerChat.receive(1000)
    assert pkt == 4

    # Should not raise
    PeerChat.send_ping("py_ping_target")

    PeerChat.disconnect()
